src/System.IO.Pipelines/src/System/IO/Pipelines/PipeOptions.cs

```diff
@@ -45,7 +45,7 @@ public class PipeOptions
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.resumeWriterThreshold);
             }
 
-            Pool = pool ?? MemoryPool<byte>.Shared;
+            Pool = pool;
```

Comment:
Should the Pool getter check whether it's null and return MemoryPool.Shared?

Reply:
And do we already have tests that exercise all the new paths?

Reply:
We should have coverage for both cases. The tests use a custom pool in some cases to verify we're renting and returning things properly. The only esoteric thing I can see failing is somebody creating a PipeOptions and accessing the Pool property will now be null instead MemoryPool<byte>.Shared but it feels like a super edge case.

Reply:
> The only esoteric thing I can see failing is somebody creating a PipeOptions and accessing the Pool property will now be null instead MemoryPool.Shared but it feels like a super edge case.

That's why I suggested above making the Pool property be:
```C#
private readonly MemoryPool<byte> _pool;
public MemoryPool<byte> Pool => _pool ?? MemoryPool<byte>.Shared;
```
:smile:

Reply:
I wouldn't want to compare against the static shared property on every rent, checking null is much faster. I could assign a null pool in the Pipe constructor if the PipeOptions gives me a MemoryPool<byte>.Shared.

Reply:
Ok.  Hopefully no one does `_options.Pool.Rent(…)` :wink:

Reply:
It's not even stored 😄  https://github.com/dotnet/corefx/blob/24e8722c1e7667ba36c2cb622f5cad480a56f2e8/src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L103

Reply:
Does someone creating their own PipeReader/Writer not use PipeOptions?

Reply:
I mean, they *can* and but they can get the optimized version. I was suggesting leaving `PipeOptions` with the `MemoryPool<byte>.Shared` and setting the private field in Pipe to null if options.Pool is shared. So it's non non-breaking and it doesn't affect other potential consumers. The ctor would look like this:

_pool  = options.Pool == MemoryPool<byte>.Shared ? null : options.Pool;

Reply:
Ah, great, works for me.  Thanks for humoring me.